src/System.Management.Automation/utils/Telemetry.cs

```diff
@@ -0,0 +1,407 @@
+        /// Static constructor determines whether telemetry is to be sent, and then
+        /// sets the telemetry key and set the telemetry delivery mode.
+        /// Creates the session ID and initializes the HashSet of known module names.
+        /// Gets or constructs the unique identifier.
+        /// </summary>
+        static ApplicationInsightsTelemetry()
+        {
+            // If we can't send telemetry, there's no reason to do any of this
+            CanSendTelemetry = !GetEnvironmentVariableAsBool(name: _telemetryOptoutEnvVar, defaultValue: false);
+            if (CanSendTelemetry)
+            {
+                s_telemetryClient = new TelemetryClient();
+                TelemetryConfiguration.Active.InstrumentationKey = _psCoreTelemetryKey;
+                TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = s_developerMode;
+                s_sessionId = Guid.NewGuid();
+
+                // use a hashset when looking for module names, it should be quicker than a string comparison
+                s_knownModules = new HashSet<string>(s_knownModuleNames, StringComparer.OrdinalIgnoreCase);
+                s_uniqueUserIdentifier = GetUniqueIdentifier();
+                s_developerMode = false;
+            }
+        }
+
+        /// <summary>
+        /// Determine whether the environment variable is set and how.
+        /// </summary>
+        /// <param name="name">The name of the environment variable.</param>
+        /// <param name="defaultValue">If the environment variable is not set, use this as the default value.</param>
+        /// <returns>A boolean representing the value of the environment variable.</returns>
+        private static bool GetEnvironmentVariableAsBool(string name, bool defaultValue)
+        {
+            var str = Environment.GetEnvironmentVariable(name);
+            if (string.IsNullOrEmpty(str))
+            {
+                return defaultValue;
+            }
+
+            switch (str.ToLowerInvariant())
+            {
+                case "true":
+                case "1":
+                case "yes":
+                    return true;
+                case "false":
+                case "0":
+                case "no":
+                    return false;
+                default:
+                    return defaultValue;
+            }
+        }
+
+        /// <summary>
+        /// Send telemetry as a metric.
+        /// </summary>
+        /// <param name="metricId">The type of telemetry that we'll be sending.</param>
+        /// <param name="data">The specific details about the telemetry.</param>
+        internal static void SendTelemetryMetric(TelemetryType metricId, string data)
+        {
+            if (!CanSendTelemetry)
+            {
+                return;
+            }
+
+            string metricName = metricId.ToString();
+            string uuidString = s_uniqueUserIdentifier.ToString();
+            string sessionIdString  = s_sessionId.ToString();
```

Comment:
I recommend to make `s_uniqueUserIdentifier` and `s_sessionId` string type, so we only call `ToString()` on the GUID once. Now we are creating new strings every time `SendTelemetryMetric` is called.

Reply:
fixed